File: app/src/main/java/online/fatbook/fatbookapp/ui/viewmodel/UserViewModel.java
package online.fatbook.fatbookapp.ui.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import online.fatbook.fatbookapp.core.Recipe;
import online.fatbook.fatbookapp.core.User;

import java.util.List;

public class UserViewModel extends ViewModel {

    private final MutableLiveData<User> user = new MutableLiveData<User>();

    private final MutableLiveData<List<Recipe>> recipeList = new MutableLiveData<>();

    private final MutableLiveData<List<Recipe>> bookmarkedRecipeList = new MutableLiveData<>();

    private final MutableLiveData<List<Recipe>> forkedRecipeList = new MutableLiveData<>();

    private final MutableLiveData<List<Recipe>> feedRecipeList = new MutableLiveData<>();

    public void setUser(User user) {
        this.user.setValue(user);
    }

    public LiveData<User> getUser() {
        return user;
    }

    public void setRecipeList(List<Recipe> recipeList) {
        this.recipeList.setValue(recipeList);
    }

    public LiveData<List<Recipe>> getRecipeList() {
        return recipeList;
    }

    public void setBookmarkedRecipeList(List<Recipe> bookmarkedRecipeList) {
        this.bookmarkedRecipeList.setValue(bookmarkedRecipeList);
    }

    public LiveData<List<Recipe>> getBookmarkedRecipeList() {
        return bookmarkedRecipeList;
    }

    public void setForkedRecipeList(List<Recipe> forkedRecipeList) {
        this.forkedRecipeList.setValue(forkedRecipeList);
    }

    public LiveData<List<Recipe>> getForkedRecipeList() {
        return forkedRecipeList;
    }

    public void setFeedRecipeList(List<Recipe> feedRecipeList) {
        this.feedRecipeList.setValue(feedRecipeList);
    }

    public LiveData<List<Recipe>> getFeedRecipeList() {
        return feedRecipeList;
    }
}

File: app/src/main/java/online/fatbook/fatbookapp/ui/viewmodel/UserViewModel.kt
package online.fatbook.fatbookapp.ui.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import online.fatbook.fatbookapp.callback.ResultCallback
import online.fatbook.fatbookapp.core.user.User
import online.fatbook.fatbookapp.network.UserUpdateRequest
import online.fatbook.fatbookapp.repository.UserRepository

class UserViewModel : ViewModel() {

    companion object {
        private const val TAG = "UserViewModel"
    }

    private val repository by lazy { UserRepository() }

    private val _selectedUsername = MutableLiveData("")
    val selectedUsername: LiveData<String> = _selectedUsername

    fun setSelectedUsername(value: String) {
        _selectedUsername.value = value
    }

    private val _user = MutableLiveData<User>()
    val user: LiveData<User> get() = _user

    fun setUser(value: User) {
        _user.value = value
    }

    private val _resultCode = MutableLiveData<Int?>()
    val resultCode: LiveData<Int?> get() = _resultCode

    fun setResultCode(value: Int?) {
        _resultCode.value = value
    }

    fun loadCurrentUser(username: String, callback: ResultCallback<User>) {
        repository.getUserByUsername(username, object : ResultCallback<User> {
            override fun onResult(value: User?) {
                setUser(value!!)
                callback.onResult(value)
            }

            override fun onFailure(value: User?) {
            }
        })
    }

    fun getUserByUsername(username: String, callback: ResultCallback<User>) {
        repository.getUserByUsername(username, object : ResultCallback<User> {
            override fun onResult(value: User?) {
                callback.onResult(value)
            }

            override fun onFailure(value: User?) {
            }
        })
    }

    fun updateUser(request: UserUpdateRequest, callback: ResultCallback<User>) {
        repository.updateUser(request, object : ResultCallback<User> {
            override fun onResult(value: User?) {
                callback.onResult(value)
            }

            override fun onFailure(value: User?) {
                callback.onFailure(value)
            }

        })
    }

}